rxtools/source/features/downgradeapp.c

```diff
@@ -11,169 +11,582 @@
 #include "TitleKeyDecrypt.h"
 #include "NandDumper.h"
 #include "aes.h"
+#include "polarssl/sha2.h"
 
-typedef struct{
+#define bswap_32(a) ((((a) << 24) & 0xff000000) | (((a) << 8) & 0xff0000) | (((a) >> 8) & 0xff00) | (((a) >> 24) & 0xff))
+
+typedef struct {
 	unsigned int id;
 	unsigned short index;
 	unsigned short type;
-	unsigned char size[8];
+	unsigned int size;
 	unsigned char signature[0x20];
-} tmd_chunck_struct;
+} tmd_chunk_struct;
+
+unsigned char region = 0;
+char *regions[6] = { "Japan", "USA", "Europe", "China", "Korea", "Taiwan" };
 
 char tmpstr[256];
-char path[256];
 FILINFO curInfo;
 DIR myDir;
 
-int FindApp(char* filepath, unsigned int tid_low, unsigned int tid_high, char* drive){
-    char* folder = &tmpstr;
+char cntpath[256]; // Contains the NAND content path
+char tmdpath[256]; // Contains the NAND TMD path
+
+void print_sha256(unsigned char hash[32])
+{
+	int i;
+	for (i = 0; i < 32; i++)
+	{
+		print("%02X", hash[i]); ConsoleShow();
```

Comment:
@DarkMatterCore Thanks. If you want, you can change & test & push, just before this is merged..